### PR TITLE
Switch Deployments to new hashing algo w/ collision avoidance mechanism

### PR DESCRIPTION
**Pull request by 0xmichalis**
Implements https://github.com/kubernetes/community/pull/477

@kubernetes/sig-apps-api-reviews @kubernetes/sig-apps-pr-reviews 

Fixes https://github.com/kubernetes/kubernetes/issues/29735
Fixes https://github.com/kubernetes/kubernetes/issues/43948

```release-note
Deployments are updated to use (1) a more stable hashing algorithm (fnv) than the previous one (adler) and (2) a hashing collision avoidance mechanism that will ensure new rollouts will not block on hashing collisions anymore.
```

### CODE REVIEW

File: pkg/apis/apps/v1beta1/types.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta1/types.go]: I think we need `+optional` here
Ref #34860, cc @mehdy

Comment (kargakis) [inline pkg/apis/apps/v1beta1/types.go]: Do we also need this tag in the internal types?

Comment (janetkuo) [inline pkg/apis/apps/v1beta1/types.go]: Yes we do

**Comment by janetkuo** [comment]
We sync status by calling `calculateStatus` first, which basically wipes out whatever is set in Deployment `status.uniquifier`

**Comment by kargakis** [comment]
> We sync status by calling calculateStatus first, which basically wipes out whatever is set in Deployment status.uniquifier

Yeah, on my list to fix.

File: pkg/controller/deployment/util/pod_util.go
Comment (kargakis) [inline pkg/controller/deployment/util/pod_util.go]: Can an int64 integer be larger than 8 bytes?

Comment (kargakis) [inline pkg/controller/deployment/util/pod_util.go]: I still need to unit test this

Comment (janetkuo) [inline pkg/controller/deployment/util/pod_util.go]: I did some experiment https://play.golang.org/p/3DKSA8wYup

Comment (kargakis) [inline pkg/controller/deployment/util/pod_util.go]: Interesting. Updated and also added a simple unit test.

**Comment by kargakis** [comment]
@janetkuo @kubernetes/sig-apps-pr-reviews this is ready for reviews. I will follow up with unit tests for getNewReplicaSet in a different PR because I want to break down that function into pieces before unit-testing it and I don't want to make reviews harder here.

**Comment by kargakis** [comment]
@lukaszo can you also have a look?

File: pkg/apis/apps/v1beta1/types.go
Comment (lukaszo) [inline pkg/apis/apps/v1beta1/types.go]: Why do we keep this in status? It seems little strange.
Also, who can update status? Only controller?

File: pkg/apis/extensions/validation/validation.go
Comment (lukaszo) [inline pkg/apis/extensions/validation/validation.go]: can you add a test for this case?

File: pkg/apis/apps/v1beta1/types.go
Comment (janetkuo) [inline pkg/apis/apps/v1beta1/types.go]: Putting it in status was discussed in the proposal for hash collision. We don't want users to set it or think about it. 

API server can update status as well (if we implement it).

**Comment by janetkuo** [comment]
Sorry, didn't have time to review this further. Taking a look now.

File: test/e2e/framework/util.go
Comment (janetkuo) [inline test/e2e/framework/util.go]: (optional) create `const` for poll interval and timeout so that we have consistent value in all `func UpdateXxxWithRetries`

File: pkg/apis/extensions/validation/validation.go
Comment (kargakis) [inline pkg/apis/extensions/validation/validation.go]: done

File: test/e2e/deployment.go
Comment (kargakis) [inline test/e2e/deployment.go]: done

File: test/e2e/framework/util.go
Comment (kargakis) [inline test/e2e/framework/util.go]: Prefer to leave this change for a follow-up

File: pkg/apis/extensions/validation/validation.go
Comment (kargakis) [inline pkg/apis/extensions/validation/validation.go]: I don't feel strong for tightening this to check that the field increases by 1 - either way if we stabilize to a value, we won't care. Added validation that it won't decrease.

**Comment by kargakis** [comment]
@lukaszo @janetkuo comments addressed - I would like to get this merged so that our QA team can get on it soon.

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
LGTM now. Make the squashes you need and I'll apply the tag.

File: pkg/apis/extensions/validation/validation.go
Comment (janetkuo) [inline pkg/apis/extensions/validation/validation.go]: My worry is that some stale Deployment status in cache may cause uniquifier to decrease or increase too much. We need the value to be stable for idempotent RS names. This can be a follow-up.

**Comment by kargakis** [comment]
@janetkuo squashed

@smarterclayton needs an approval

**Comment by janetkuo** [comment]
```
FAILED   hack/make-rules/../../hack/verify-federation-openapi-spec.sh
```

**Comment by kargakis** [comment]
@k8s-bot gce etcd3 e2e test this up failure

**Comment by smarterclayton** [comment]
@kubernetes/api-reviewers any concerns about the use of `Uniquifier` as the field name?  It's somewhat unusual in our lexicon.

**Comment by kargakis** [comment]
Rebased on top of conflicts in generated code, re-applying lgtm

**Comment by kargakis** [comment]
> @kubernetes/api-reviewers any concerns about the use of Uniquifier as the field name? It's somewhat unusual in our lexicon.

Unless there is something else we can use, I am fine with moving forward with this one. It's not a field that most users need to know anyway.

**Comment by kargakis** [comment]
@kubernetes/test-infra-maintainers was there a change wrt the rebase label recently in the respective bot? I don't see its history in the PR anymore.

**Comment by kargakis** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot kops aws e2e test this
@k8s-bot gce etcd3 e2e test this

**Comment by kargakis** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by janetkuo** [comment]
Looks like there's no objection. Shall we go ahead and merge this?

**Comment by kargakis** [comment]
One more rebase - reapplying the label

**Comment by kargakis** [comment]
@kubernetes/sig-apps-api-reviews any comments on the name or anything else?

**Comment by kargakis** [comment]
@smarterclayton re the name, how about any of `collisionIndex`, `collisionCounter`, `hashingIndex`?

**Comment by kargakis** [comment]
@k8s-bot unit test this https://github.com/kubernetes/kubernetes/issues/46204
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot kops aws e2e test this

**Comment by smarterclayton** [comment]
How about `collisionCounter`?  That's more accurate and easier for a new user to understand?

**Comment by kargakis** [comment]
SGTM - @janetkuo got any objections for `collisionCounter` instead of `uniquifier`?

**Comment by kargakis** [comment]
> SGTM - @janetkuo got any objections for collisionCounter instead of uniquifier?

@kubernetes/sig-apps-api-reviews

**Comment by kargakis** [comment]
I will add the appropriate labels at 12:00 CET tomorrow if there are no comments until then.

**Comment by kargakis** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by mikedanese** [comment]
I've seen "uniquifier" referred to as "fingerprint" in other apis. Is there significance to the monotonically incrementing or is that an implementation detail? Is it guaranteed dense?

File: pkg/apis/apps/v1beta1/types.go
Comment (mikedanese) [inline pkg/apis/apps/v1beta1/types.go]: Is it significant to note that this is an upper bound on collision, not the actual number of collisions?

**Comment by kargakis** [comment]
> I've seen "uniquifier" referred to as "fingerprint" in other apis. Is there significance to the monotonically incrementing or is that an implementation detail? Is it guaranteed dense?

There is significance in that it's incrementing so in case of a collision we are essenitally changing the algo used to compute a hash and it should be a new one. It is used as a counter of the actual collisions.

**Comment by wanghaoran1988** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by lavalamp** [comment]
I vote for `collisionCounter` and against `uniquifier` which is super unintuitive.

`revision`, `generation`, `historyIndex` all are better. `fingerprint` and `etag` both seem like the wrong concept.

**Comment by janetkuo** [comment]
`collisionCounter` is fine. `revision`, `generation`, `historyIndex` aren't appropriate because this is not a field for version-like information but just a way to avoid name collisions when hash collision happens.

**Comment by bgrant0607** [comment]
@kargakis 

PSA: I don't receive ANY github notifications now. If something needs my attention, please resort to other means, such as slack or email.

Clayton asked for my opinion via slack. I said I was ok with `collisionCount`(er).

The fact that it is sequential is not really material, nor strongly guaranteed, since there are scenarios where it could double count or reset, but "count" makes it easy to understand what it's doing. I oppose any name containing "index", "revision", or "generation", that might imply stronger semantics, be confused with other fields of the same names, or otherwise lead to an expectation that the value would appear in labels, annotations, or resource names.

**Comment by bgrant0607** [comment]
Please remove the word "monotonically" from the API field descriptions.

**Comment by bgrant0607** [comment]
Also: @ncdc suggested hashCollisions, which isn't bad, though I'm not 100% committed to the idea of keeping a hash forever. Just `collisions` is a bit too obscure. `resourceCollisions` could potentially work, but `collisionCount` is good enough to call it done and move on.

**Comment by thockin** [comment]
I'm concerned that this is kept in Status, but doesn't really pass the litmus test of "could be rebuilt by observation",

**Comment by kargakis** [comment]
> I'm concerned that this is kept in Status, but doesn't really pass the litmus test of "could be rebuilt by observation",

I see your point and we could potentially annotate ReplicaSets with the collisionCounter they were created and reflect the highest adopted value back to the Deployment status but eventually this specific field doesn't need to be rebuilt, mostly due to our defaults (new Deployments in the apps group are defaulted to retaining 3 old ReplicaSets, coupled with fnv it's unlike that we will ever have a collision again).

**Comment by kargakis** [comment]
Updated to use `collisionCount` and fixed the API comment.

**Comment by kargakis** [comment]
Approving assuming consensus on the name. Adding lgtm based on prior lgtm.

**Pull request merged**
